File: packages/python/src/openskp/parser.py
"""TLV binary parser for SketchUp's internal data format.

This module provides low-level functions for reading the Tag-Length-Value (TLV)
encoded binary data found inside SketchUp files. The parser recursively
descends into container tags to build a tree of ``TlvNode`` objects.

Typical usage::

    from openskp.parser import parse_tlv_recursive
    from openskp.model import TlvNode

    nodes = parse_tlv_recursive(raw_bytes, 0, len(raw_bytes))
"""

from __future__ import annotations

import struct
from typing import List, Optional, Set

from .model import TlvNode

# ---------------------------------------------------------------------------
# Known container tags that may hold nested TLV children.
# This set is used by ``parse_tlv_recursive`` to decide whether to descend
# into a tag's payload.
# ---------------------------------------------------------------------------
CONTAINER_TAGS: Set[str] = {
    'F401', 'F701', 'D430', 'D530', 'C832',
    '7C15', '8813', '8913', '8A13', '8B13', '8C13', '8D13', '4C1D', '6419',
    'F901', '7017', '7117', 'D007', 'C409', '9411', '9511', '0F01',
    '384A', 'B80B', '9713', '2C4C', 'AC0D', 'AE0D', 'F601', 'F801',
    '983A', '993A', '8C3C', '8D3C',
}


# ── Primitive readers ──────────────────────────────────────────────────────


def read_u32(data: bytes, offset: int) -> int:
    """Read an unsigned 32-bit little-endian integer.

    Args:
        data: Raw byte buffer.
        offset: Byte offset into *data*.

    Returns:
        The decoded ``uint32`` value.

    Raises:
        struct.error: If there are fewer than 4 bytes at *offset*.
    """
    return struct.unpack_from('<I', data, offset)[0]


def read_f64(data: bytes, offset: int) -> float:
    """Read a 64-bit little-endian IEEE-754 double.

    Args:
        data: Raw byte buffer.
        offset: Byte offset into *data*.

    Returns:
        The decoded ``float64`` value.

    Raises:
        struct.error: If there are fewer than 8 bytes at *offset*.
    """
    return struct.unpack_from('<d', data, offset)[0]


def parse_var_int(data: bytes, offset: int, length: int) -> int:
    """Decode a variable-length little-endian unsigned integer.

    Args:
        data: Raw byte buffer.
        offset: Byte offset into *data*.
        length: Number of bytes to consume (1–8).

    Returns:
        The decoded integer value.
    """
    val: int = 0
    for i in range(length):
        val |= data[offset + i] << (8 * i)
    return val


# ── Recursive TLV parser ──────────────────────────────────────────────────


def parse_tlv_recursive(
    data: bytes,
    start: int,
    end: int,
    container_tags: Optional[Set[str]] = None,
    depth: int = 0,
) -> List[TlvNode]:
    """Recursively parse a TLV-encoded byte range into a tree of nodes.

    Each TLV element consists of:

    * **Tag** — 2 bytes (interpreted as upper-case hex, e.g. ``"FF01"``).
    * **Length** — 4-byte ``uint32`` little-endian payload length.
    * **Value** — *length* bytes of payload data.

    If a tag is listed in *container_tags* the parser descends into its
    payload to extract child nodes.

    Args:
        data: Full binary buffer.
        start: Byte offset where parsing begins (inclusive).
        end: Byte offset where parsing stops (exclusive).
        container_tags: Set of hex tag strings that may contain nested TLV
            data.  Defaults to :data:`CONTAINER_TAGS`.
        depth: Current recursion depth (used internally).

    Returns:
        A list of :class:`TlvNode` objects found in the range
        ``[start, end)``.
    """
    if container_tags is None:
        container_tags = CONTAINER_TAGS

    pos: int = start
    elements: List[TlvNode] = []

    while pos <= end - 6:
        tag_bytes = data[pos:pos + 2]
        size = read_u32(data, pos + 2)

        if pos + 6 + size > end:
            break

        tag_hex: str = tag_bytes.hex().upper()
        children: List[TlvNode] = []
        is_container: bool = tag_hex in container_tags

        if is_container and size > 0:
            children = parse_tlv_recursive(
                data, pos + 6, pos + 6 + size, container_tags, depth + 1
            )

        payload: bytes = data[pos + 6: pos + 6 + size] if not children else b''

        elements.append(TlvNode(
            offset=pos,
            tag=tag_hex,
            size=size,
            children=children,
            payload=payload,
        ))

        pos += 6 + size

    return elements

File: packages/typescript/src/parser.ts
export interface TlvNode {
  offset: number;
  tag: string;
  size: number;
  children: TlvNode[];
  payload: Uint8Array;
}

/**
 * VFF (2021+) model.dat binary structure uses Type-Length-Value (TLV) records.
 * Most TLV tags carry raw leaf binary/string payloads, but container tags contain
 * nested sub-TLV nodes (e.g. definitions, drawing elements, component instances).
 * CONTAINER_TAGS lists every tag hex ID that the TLV parser must recursively traverse.
 */
export const CONTAINER_TAGS = new Set<string>([
  'F401', 'F701', 'D430', 'D530', 'C832',
  '7C15', '8813', '8913', '8A13', '8B13', '8C13', '8D13', '4C1D', '6419',
  'F901', '7017', '7117', 'D007', 'C409', '9411', '9511', '0F01',
  '384A', 'B80B', '9713', '2C4C', 'AC0D', 'AE0D', 'F601', 'F801',
  '983A', '993A', '8C3C', '8D3C',
  // Image-entity placement: an Image placed in the model wraps a standard
  // 6419 instance node inside 9013 -> 401F. Without these two containers,
  // that inner instance stays buried in an opaque payload and the image
  // definition looks "never placed".
  '9013', '401F',
]);

export function readU32(data: Uint8Array, offset: number): number {
  if (offset + 4 > data.length) {
    throw new Error('Out of bounds readU32');
  }
  const view = new DataView(data.buffer, data.byteOffset + offset, 4);
  return view.getUint32(0, true);
}

export function readF64(data: Uint8Array, offset: number): number {
  if (offset + 8 > data.length) {
    throw new Error('Out of bounds readF64');
  }
  const view = new DataView(data.buffer, data.byteOffset + offset, 8);
  return view.getFloat64(0, true);
}

export function readI32(data: Uint8Array, offset: number): number {
  if (offset + 4 > data.length) {
    throw new Error('Out of bounds readI32');
  }
  const view = new DataView(data.buffer, data.byteOffset + offset, 4);
  return view.getInt32(0, true);
}

export function parseVarInt(data: Uint8Array, offset: number, length: number): number {
  let val = 0;
  for (let i = 0; i < length; i++) {
    val += data[offset + i] * Math.pow(256, i);
  }
  return val;
}

export function parseTlvRecursive(
  data: Uint8Array,
  start: number,
  end: number,
  containerTags: Set<string> = CONTAINER_TAGS,
  depth: number = 0
): TlvNode[] {
  let pos = start;
  const elements: TlvNode[] = [];

  while (pos <= end - 6) {
    const tagBytes = data.subarray(pos, pos + 2);
    const size = readU32(data, pos + 2);

    if (pos + 6 + size > end) {
      break;
    }

    // Convert tagBytes to uppercase hex string
    let tagHex = '';
    for (let i = 0; i < 2; i++) {
      const hex = tagBytes[i].toString(16).toUpperCase();
      tagHex += hex.length === 1 ? '0' + hex : hex;
    }

    let children: TlvNode[] = [];
    const isContainer = containerTags.has(tagHex);

    if (isContainer && size > 0) {
      children = parseTlvRecursive(
        data,
        pos + 6,
        pos + 6 + size,
        containerTags,
        depth + 1
      );
    }

    const payload = children.length > 0 ? new Uint8Array(0) : data.subarray(pos + 6, pos + 6 + size);

    elements.push({
      offset: pos,
      tag: tagHex,
      size,
      children,
      payload,
    });

    pos += 6 + size;
  }

  return elements;
}

/** Scan [start, end) for direct-child (tag, offset, size) headers only,
 * without recursing into any container - O(sibling count), not O(total
 * node count). Used by iterTopLevelLazy to locate top-level records one at
 * a time. */
function flatHeaders(data: Uint8Array, start: number, end: number): [string, number, number][] {
  const headers: [string, number, number][] = [];
  let pos = start;
  while (pos <= end - 6) {
    const tagBytes = data.subarray(pos, pos + 2);
    const size = readU32(data, pos + 2);
    if (pos + 6 + size > end) break;
    let tagHex = '';
    for (let i = 0; i < 2; i++) {
      const hex = tagBytes[i].toString(16).toUpperCase();
      tagHex += hex.length === 1 ? '0' + hex : hex;
    }
    headers.push([tagHex, pos, size]);
    pos += 6 + size;
  }
  return headers;
}

export interface TopLevelRecord {
  index: number;
  total: number;
  node: TlvNode;
}

/** Yield `{index, total, node}` for each top-level TLV record's
 * fully-recursed node one at a time, transparently unwrapping a lone "F401"
 * wrapper - without ever materializing more than one top-level subtree
 * simultaneously. `total` (the top-level sibling count) comes for free from
 * the same cheap header scan that drives the loop, so callers can report
 * "N of total" progress with no extra pass over the file. Each yielded node
 * is safe to discard (drop all references) once the caller is done with it,
 * before the next one is produced - that's what keeps peak memory bounded
 * by the size of the single largest top-level record instead of the whole
 * file (real production files can have 100k+ separate definitions). */
export function* iterTopLevelLazy(
  data: Uint8Array,
  start: number,
  end: number,
  containerTags: Set<string> = CONTAINER_TAGS
): Generator<TopLevelRecord> {
  let headers = flatHeaders(data, start, end);
  if (headers.length === 1 && headers[0][0] === 'F401') {
    const [, f401Offset, f401Size] = headers[0];
    headers = flatHeaders(data, f401Offset + 6, f401Offset + 6 + f401Size);
  }

  const total = headers.length;
  let index = 0;
  for (const [, offset, size] of headers) {
    const recordEnd = offset + 6 + size;
    const nodes = parseTlvRecursive(data, offset, recordEnd, containerTags);
    if (nodes.length > 0) {
      yield { index, total, node: nodes[0] };
    }
    index++;
  }
}
